aws/credentials/processcreds/provider.go

```diff
@@ -0,0 +1,421 @@
+
+	// ErrMsgProcessProviderTimeout process took longer than allowed
+	ErrMsgProcessProviderTimeout = "credential process timed out"
+
+	// ErrMsgProcessProviderNoKill process could not be killed
+	ErrMsgProcessProviderNoKill = "unable to kill process"
+
+	// ErrMsgProcessProviderProcess process error
+	ErrMsgProcessProviderProcess = "error in credential_process"
+
+	// ErrMsgProcessProviderParse problem parsing output
+	ErrMsgProcessProviderParse = "parse failed of credential_process output"
+
+	// ErrMsgProcessProviderVersion version error in output
+	ErrMsgProcessProviderVersion = "wrong version in process output (not 1)"
+
+	// ErrMsgProcessProviderMissKey missing access key id in output
+	ErrMsgProcessProviderMissKey = "missing AccessKeyId in process output"
+
+	// ErrMsgProcessProviderMissSecret missing secret acess key in output
+	ErrMsgProcessProviderMissSecret = "missing SecretAccessKey in process output"
+
+	// ErrMsgProcessProviderPrepareCmd prepare of command failed
+	ErrMsgProcessProviderPrepareCmd = "failed to prepare command"
+
+	// ErrMsgProcessProviderFewArgs not enough args
+	ErrMsgProcessProviderFewArgs = "not enough args"
+
+	// ErrMsgProcessProviderEmptyCmd command must not be empty
+	ErrMsgProcessProviderEmptyCmd = "command must not be empty"
+
+	// ErrMsgProcessProviderPipe failed to initialize pipe
+	ErrMsgProcessProviderPipe = "failed to initialize pipe"
+
+	// GoOSWindowsKey for identifying Windows OS
+	GoOSWindowsKey = "windows"
+
+	// DefaultDuration is the default amount of time in minutes that the
+	// credentials will be valid for.
+	DefaultDuration = time.Duration(15) * time.Minute
+
+	// DefaultInitialBufSize size for initial buffer.
+	DefaultInitialBufSize = 200
+
+	// DefaultMaxBufSize limits buffer size from growing to an enormous
+	// amount due to a faulty process.
+	DefaultMaxBufSize = 512
+
+	// DefaultTimeout default limit on time a process can run.
+	DefaultTimeout = time.Duration(1) * time.Minute
+)
+
+// ProcessProvider satisfies the credentials.Provider interface, and is a
+// client to retrieve credentials from a process.
+type ProcessProvider struct {
+	staticCreds bool
+	credentials.Expiry
+	originalCommand []string
+
+	// Expiry duration of the credentials. Defaults to 15 minutes if not set.
+	Duration time.Duration
+
+	// ExpiryWindow will allow the credentials to trigger refreshing prior to
+	// the credentials actually expiring. This is beneficial so race conditions
+	// with expiring credentials do not cause request to fail unexpectedly
+	// due to ExpiredTokenException exceptions.
+	//
+	// So a ExpiryWindow of 10s would cause calls to IsExpired() to return true
+	// 10 seconds before the credentials are actually expired.
+	//
+	// If ExpiryWindow is 0 or less it will be ignored.
+	ExpiryWindow time.Duration
+
+	// A string representing an os command that should return a JSON with
+	// credential information.
+	Command *exec.Cmd
```

Comment:
Having the command publicly accessible i think is risky. If a user constructs a command with additional configuration set that the process credential provider isn't aware of they could receive unexpected behavior the second+ time the credentials are retrieved.

Could you change this member to not be exported.

Reply:
Yes. This necessitated a new constructor (and test to cover the constructor).